robotExecution/socketConnection.js

```diff
@@ -40,17 +42,31 @@ exports.connectWithSocket = (configurationData) => {
       }
     );
     // test.robot is just used for testing purposes. Use executable.robot for the real product.
-    exec('robot ./robotExecution/executable.robot', (err) => {
-      if (err) {
-        console.log(err.message);
-        socket.emit('updatedRobotJob', { jobId, status: 'failed' });
+    exec(
+      'robot --listener ./robotMonitoring/LiveLogsListener.py ./robotExecution/executable.robot',
+      (err) => {
+        if (err) {
+          console.log(err.message);
+        }
+        // delete all unnecessary files after robot execution
+        exec('find . -name "*.xml" -type f|xargs rm -f');
+        exec('find . -name "*.html" -type f|xargs rm -f');
+        exec('find . -name "*.log" -type f|xargs rm -f');
       }
-      socket.emit('updatedRobotJobStatus', { jobId, status: 'success' });
-
-      // delete all unnecessary files after robot execution
-      exec('find . -name "*.xml" -type f|xargs rm -f');
-      exec('find . -name "*.html" -type f|xargs rm -f');
-      exec('find . -name "*.log" -type f|xargs rm -f');
+    );
+    const watcher = chokidar.watch('./robotMonitoring/robotLogs.json');
+    watcher.on('change', () => {
+      let robotLogs = getRobotLogs();
+      while (true) {
```

Comment:
I really do not like this while Loop even after Daniel explained it to me. You should consider building a while-loop that has a timeout, a time after which it will automatically stop retrying and notify the user through an error or so

Reply:
I might have found a solution to evade this loop using chokidars awaitWriteFinish option:

```javascript 
const watcher = chokidar.watch('./robotMonitoring/robotLogs.json', {
      awaitWriteFinish: { stabilityThreshold: 500 },
    });
    watcher.on('change', () => {
      let robotLogs = getRobotLogs();
      try {
        robotLogs = JSON.parse(robotLogs);
        socket.emit('updatedRobotJob', { jobId, robotLogs });
      } catch (error) {
        /*  robotLogs = getRobotLogs();
        continue; */
        console.log(error);
      }
    }
```

Source: https://github.com/paulmillr/chokidar#performance

> `awaitWriteFinish` (default: `false`).
> By default, the `add` event will fire when a file first appears on disk, before
> the entire file has been written. Furthermore, in some cases some `change`
> events will be emitted while the file is being written. In some cases,
> especially when watching for large files there will be a need to wait for the
> write operation to finish before responding to a file creation or modification.
> Setting `awaitWriteFinish` to `true` (or a truthy value) will poll file size,
> holding its `add` and `change` events until the size does not change for a
> configurable amount of time. The appropriate duration setting is heavily
> dependent on the OS and hardware. For accurate detection this parameter should
> be relatively high, making file watching much less responsive.
> Use with caution.
>   * *`options.awaitWriteFinish` can be set to an object in order to adjust
>   timing params:*
>   * `awaitWriteFinish.stabilityThreshold` (default: 2000). Amount of time in
>   milliseconds for a file size to remain constant before emitting its event.
>   * `awaitWriteFinish.pollInterval` (default: 100). File size polling interval, in milliseconds.

Reply:
I think the proposed solution by Sandro is great. I just implemented and tested it on my local machine and it works 👍 This solution would also make the robotLogsHelper.js obsolete. If you also like the proposed solution @kej-jay then we will implement it.